server: TestStatusEngineStatsJson failed

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9195874?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9195874?buildTab=artifacts#/) on release-23.1 @ [dc054b59d684e421238d7044539463df051186b7](https://github.com/cockroachdb/cockroach/commits/dc054b59d684e421238d7044539463df051186b7):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/74885ac1e7f972ae9e97608ed3fdb458/logTestStatusEngineStatsJson1821925975
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:282: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/74885ac1e7f972ae9e97608ed3fdb458/logTestStatusEngineStatsJson1821925975
--- FAIL: TestStatusEngineStatsJson (15.37s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #95491 server: TestStatusEngineStatsJson failed [A-observability-inf C-test-failure O-robot T-observability-inf branch-master]
</p>
</details>
/cc @cockroachdb/obs-inf-prs @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25791
Epic: CRDB-28893

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9698742?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9698742?buildTab=artifacts#/) on master @ [568c68a5b2ba48ba08cd69c96876dd8ac25ade3c](https://github.com/cockroachdb/cockroach/commits/568c68a5b2ba48ba08cd69c96876dd8ac25ade3c):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/38c80cbed3f79a398570824e9fcc636b/logTestStatusEngineStatsJson2636581669
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:279: using admin URL https://127.0.0.1:44241
    status_test.go:286: condition failed to evaluate within 45s: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/38c80cbed3f79a398570824e9fcc636b/logTestStatusEngineStatsJson2636581669
--- FAIL: TestStatusEngineStatsJson (84.09s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9846375?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9846375?buildTab=artifacts#/) on master @ [ddc247fd1fa4c7c1d2f212fcdd101ef200b772ff](https://github.com/cockroachdb/cockroach/commits/ddc247fd1fa4c7c1d2f212fcdd101ef200b772ff):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/04ff595e8b8a4cdf61532aa9937d7edf/logTestStatusEngineStatsJson2340365415
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:279: using admin URL https://127.0.0.1:43193
    status_test.go:286: condition failed to evaluate within 45s: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/04ff595e8b8a4cdf61532aa9937d7edf/logTestStatusEngineStatsJson2340365415
--- FAIL: TestStatusEngineStatsJson (109.60s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9860794?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9860794?buildTab=artifacts#/) on release-23.1 @ [1f41dc622e53e7f2dcf27d80ed808a47605510b8](https://github.com/cockroachdb/cockroach/commits/1f41dc622e53e7f2dcf27d80ed808a47605510b8):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/38c80cbed3f79a398570824e9fcc636b/logTestStatusEngineStatsJson2954868437
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:284: condition failed to evaluate within 45s: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/38c80cbed3f79a398570824e9fcc636b/logTestStatusEngineStatsJson2954868437
--- FAIL: TestStatusEngineStatsJson (100.05s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9962894?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9962894?buildTab=artifacts#/) on master @ [5d775e992ade315becd6b4fc41ed1fb85d8e35d6](https://github.com/cockroachdb/cockroach/commits/5d775e992ade315becd6b4fc41ed1fb85d8e35d6):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/115b56b7b7bd2dfb6f4ea1f5f6513edb/logTestStatusEngineStatsJson1398007110
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:279: using admin URL https://127.0.0.1:35639
    status_test.go:286: condition failed to evaluate within 45s: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/115b56b7b7bd2dfb6f4ea1f5f6513edb/logTestStatusEngineStatsJson1398007110
--- FAIL: TestStatusEngineStatsJson (56.25s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9990235?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9990235?buildTab=artifacts#/) on master @ [439c515b2a0058648731da73993c409544404da1](https://github.com/cockroachdb/cockroach/commits/439c515b2a0058648731da73993c409544404da1):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestStatusEngineStatsJson483250565
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:280: using admin URL https://127.0.0.1:41589
    status_test.go:287: condition failed to evaluate within 45s: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestStatusEngineStatsJson483250565
--- FAIL: TestStatusEngineStatsJson (73.18s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10021581?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10021581?buildTab=artifacts#/) on master @ [09ec0ee7c658bc139adab023fca8a61f1cd6789f](https://github.com/cockroachdb/cockroach/commits/09ec0ee7c658bc139adab023fca8a61f1cd6789f):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestStatusEngineStatsJson835366702
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:280: using admin URL https://127.0.0.1:36079
    status_test.go:287: condition failed to evaluate within 45s: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestStatusEngineStatsJson835366702
--- FAIL: TestStatusEngineStatsJson (73.25s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/10129153?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/10129153?buildTab=artifacts#/) on master @ [76078816b69ee6d1f8dacd581ef464660fb1b697](https://github.com/cockroachdb/cockroach/commits/76078816b69ee6d1f8dacd581ef464660fb1b697):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/115b56b7b7bd2dfb6f4ea1f5f6513edb/logTestStatusEngineStatsJson3831849725
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:280: using admin URL https://127.0.0.1:35511
    status_test.go:287: condition failed to evaluate within 45s: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/115b56b7b7bd2dfb6f4ea1f5f6513edb/logTestStatusEngineStatsJson3831849725
--- FAIL: TestStatusEngineStatsJson (64.35s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestStatusEngineStatsJson [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10196078?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10196078?buildTab=artifacts#/) on release-23.1 @ [128261565cccf92b771975db161ceda9340a659a](https://github.com/cockroachdb/cockroach/commits/128261565cccf92b771975db161ceda9340a659a):


```
=== RUN   TestStatusEngineStatsJson
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestStatusEngineStatsJson1794869553
    test_log_scope.go:79: use -show-logs to present logs inline
    status_test.go:284: condition failed to evaluate within 45s: status: 503 Service Unavailable, content-type: application/json, body: {
          "error": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "code": 14,
          "message": "connection error: desc = \"transport: error while dialing: gRPC connection unexpectedly re-dialed: connection interrupted (did the remote node shut down or are there networking issues?)\"",
          "details": [
          ]
        }, error: <nil>
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/dbd7076bfe20e3183065a44fa4f7e406/logTestStatusEngineStatsJson1794869553
--- FAIL: TestStatusEngineStatsJson (62.62s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatusEngineStatsJson.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -450,6 +450,12 @@ type Context struct {
 
 // SetLoopbackDialer configures the loopback dialer function.
 func (c *Context) SetLoopbackDialer(loopbackDialFn func(context.Context) (net.Conn, error)) {
+	if c.ContextOptions.Knobs.NoLoopbackDialer {
+		// A test has decided it is opting out of the special loopback
+		// dialing mechanism. Obey it. We already have defined
+		// loopbackDialFn in that case in NewContext().
+		return
+	}
 	c.loopbackDialFn = loopbackDialFn
 }
 
@@ -1640,7 +1646,12 @@ const (
 func (rpcCtx *Context) GRPCDialOptions(
 	ctx context.Context, target string, class ConnectionClass,
 ) ([]grpc.DialOption, error) {
-	return rpcCtx.grpcDialOptionsInternal(ctx, target, class, tcpTransport)
+	transport := tcpTransport
+	if rpcCtx.Config.AdvertiseAddr == target && !rpcCtx.ClientOnly {
+		// See the explanation on loopbackDialFn for an explanation about this.
+		transport = loopbackTransport
+	}
+	return rpcCtx.grpcDialOptionsInternal(ctx, target, class, transport)
 }
 
 // grpcDialOptions produces dial options suitable for connecting to the given target and class.

--- a/pkg/server/connectivity_test.go
+++ b/pkg/server/connectivity_test.go
@@ -125,6 +125,17 @@ func TestClusterConnectivity(t *testing.T) {
 		// We're going to manually control initialization in this test.
 		NoAutoInitializeCluster: true,
 		StoreSpecs:              []base.StoreSpec{{InMemory: true}},
+		Knobs: base.TestingKnobs{
+			Server: &server.TestingKnobs{
+				ContextTestingKnobs: rpc.ContextTestingKnobs{
+					// Disable the special RPC loopback dial logic.
+					// This is needed because the test starts to perform
+					// RPC dials before the PreStart() function is called,
+					// and that is where the loopback dial function is defined.
+					NoLoopbackDialer: true,
+				},
+			},
+		},
 	}
 
 	for i, test := range testConfigurations {

